x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/build_alert.ts

```diff
@@ -179,7 +180,7 @@ export const buildAlert = (
   });
 
   return {
-    [TIMESTAMP]: new Date().toISOString(),
+    [TIMESTAMP]: alertTimestampOverride?.toISOString() ?? new Date().toISOString(),
```

Comment:
@e40pud @YulNaumenko It looks like this PR will set the timestamp of the alert document to the `task.startedAt` only in preview mode. We had discussed always setting it to the `task.startedAt` so that all alerts generated from a rule execution have the same value, instead of having slight differences based on when the alert is created. Would it be worth doing that here regardless of whether this is a preview? 

Correct me if I'm wrong, but it seems like the actual fix for this issue is in ` x-pack/plugins/security_solution/server/search_strategy/security_solution/factory/matrix_histogram/preview/query.preview_histogram.dsl.ts` which queries the `@timestamp` of the document instead of the original time?

Reply:
Correct, the actual fix is in both this and mentioned files. Here we make sure the timestamp is set to the execution start times and in the mentioned file we retrieve the correct field in case of preview histogram.

Regarding the usage of the `task.startedAt` as a timestamp for the alerts in all cases, I don't have strong opinion on that one. We do that in case of the preview to make sure that all alerts results are falling into the specified preview time frame. Otherwise, the histogram won't display the alerts which created from the events outside of the time frame (this can happen in case of the long auto-lookback time).

@marshallmain do you have more input?

Reply:
IMO, setting the `@timestamp` to the `task.startedAt` would make sense for all alerts (and would align with how the lifecycle alerts-as-data documents set `@timestamp`). It would also remove the need for the `isPreview` flag into the executors. There is some concern with introducing this new flag at this point and whether it may clash with any framework level preview functionality that may come in the future (although admittedly this is not a priority right now for Response ops).

Reply:
@marshallmain That concern about "old" timestamps is true for any rule type, including possibly long running metrics threshold rules and I don't believe there have been any users concerns raised about where the alerts appear in the table :) 

> In general I think it's important for rules to know what type of execution they're doing
We're actually leaning the opposite way, where it feels that the executor should be agnostic to whether it is a preview or not and providing this flag provides a way for all executors to possibly fork behavior based on whether `isPreview` is true, which might be harder to walk back later.

cc @mikecote @kobelb

Reply:
>In general I think it's important for rules to know what type of execution they're doing (scheduled through task manager, preview, potentially in the future ad-hoc rule runs covering old time ranges) in order to handle edge cases like this where we need slightly different behavior in different execution contexts.

In the future, the framework will play with the executor parameters (ex: from/to time range) to handle these cases (generate previews, gap coverage, etc.) so all 40+ rule types can benefit from these new platform features without changing their code.

Some examples include
- Generating previews by not persisting alert documents (or somewhere else)
- Perform gap coverage by playing with the to/from timestamps and maybe persist the documents in a special way
- Perform a test run by not persisting the outcome
 
We've already delivered one of these which is to run any rule right away "ad hoc". With these reasons, we feel giving developers more opportunities to fork code will complicate any future changes done at the platform level because we risk introducing bugs and have rules run in unexpected ways.

We've documented best practices for new rule type development (https://docs.elastic.dev/reops/developer-guide/best-practices-rules). Part of it, we highly recommend to have rule executors idempotent. By doing so, we are able to move faster as a platform to deliver product features. We still need to move towards these best practices on the existing rule types.

Reply:
I'd be happy to walk you through some of the thinking behind the best practices, discuss a future vision on how the security solution rules can focus on finding alerts and letting the platform perform the logic on them. And find a short term solution to this issue.

Reply:
My views are aligned with @mikecote and @ymao1. I do not think we should be providing the rule an `isPreview` flag for the numerous reasons they stated.

Reply:
If we remove the `isPreview` param from the global `RuleExecutorOptions` and implement the logic as part of the Security-specific preview_rules_route would that satisfy your concerns? This way there would be no changes to the alerting framework or rule registry as a result of this PR.

Reply:
Updated PR. `isPreview` parameter will only apply to security rules and be passed through `createSecurityRuleTypeWrapper`.

Reply:
If you are ok dropping functionality related to `isPreview: true` down the line, that should be ok.. So when the platform API provides a preview, ad-hoc run, etc your rules will behave like a normal run again.

Reply:
Sure, we will be happy to switch to platform API as soon as it is ready!